tutorial/practicals/LFRic/building_code/4_psydata/README.md

```diff
@@ -0,0 +1,227 @@
+values after the kernel (``perturbation_post``).
+
+At this stage no driver is created that can read in the file.
+
+
+## Step 7: Instrument more than one invoke
+
+The transformation script can be slightly changed to instrument all invokes
+in a file. While this might not be useful for kernel extraction, it is
+essential for e.g. parameter verification, where you typically want to
+check all invokes contained in a program. Use the template ``transform_all.py``
+to apply the extraction transformation to all invokes. This script actually
+requires less changes than the ``transform_one.py`` template (since it
+works on all )
+Following the same process, just using this more general transformation script,
+will result in a binary that creates two NetCDF files at run time.
+
+
+## Step 8: Try other PSyData libraries
+The following set of PSyData libraries is available and can be tested.
+Corresponding Makefiles are provided.
+
+### Readonly Verification
+This library makes sure that kernel arguments that are declared as read-only (in metadata) are not modified. The compiler
+should make sure that a variable declared with ``intent(in)`` is not explicitly
+overwritten, but a memory overwrite because of an out-of-bound access to a
+different array is always possible. To import the transformation use:
+
+```python
+    from psyclone.psyir.transformations import ReadOnlyVerifyTrans
+```
+
+Then create a transformation script based on ``transform_all.py`` script to
+instrument your application. If you run this script, nothing seems to happen.
+To see that the verification is actually working, set
+the environment variable ``PSYDATA_VERBOSE`` to either 1 or 2 - the latter
+will provide more output including each variable that is checked:
+
+```bash
+    PSYDATA_VERBOSE=2 ./time_evolution
+```
+
+If you are really daring you can modify ``prop_perturbation_kernel_mod.f90``.
+It contains some commented out code at the bottom that will overwrite
+a read-only field by using out-of-bounds array accesses (obviously you
+need make sure the compiler is not doing array bounds checking). After
+recompiling and running (even without setting ``PSYDATA_VERBOSE``), you will
+see:
+
+    20201120203040.272+1100:INFO : time_evolution_alg_step: Propagating perturbation field at timestep 1
+     ------------- PSyData -------------------------
+     Double precision field chi(1) has been modified in time_evolution : invoke_propagate_perturbation
+     Original checksum:   4657399313863802880
+     New checksum:       -5811961289819291648
+     ------------- PSyData -------------------------
+    20201120203040.286+1100:INFO : Min/max perturbation =   0.00000000E+00  0.41618197E+04
+
+Note that this error is only printed in the first time step, since after the first modification
+the modified value is not changed again.
+
+### NAN Verification
+This library verifies that all input- and output-parameters of a kernel are
```

Comment:
Again, mention the use of the appropriate makefile.

Reply:
Done.